strawberryfields/program.py

```diff
@@ -0,0 +1,897 @@
+
+
+class RegRefTransform:
+    """Represents a scalar function of one or more register references.
+
+    A RegRefTransform instance, given as a parameter to a
+    :class:`~strawberryfields.ops.Operation` constructor, represents
+    a dependence of the Operation on classical information obtained by
+    measuring one or more subsystems.
+
+    Used for deferred measurements, i.e., using a measurement's value
+    symbolically in defining a gate before the numeric value of that
+    measurement is available.
+
+    Args:
+        r (Sequence[RegRef]): register references that act as parameters for the function
+        func (None, function): Scalar function that takes the values of the
+            register references in r as parameters. None is equivalent to the identity
+            transformation lambda x: x.
+        func_str (str): an optional argument containing the string representation of the function.
+            This is useful if a lambda function is passed to the RegRefTransform, which would otherwise
+            show in the program queue as ``RegRefTransform(q[0], <lambda>)``.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, refs, func=None, func_str=None):
+        # into a list of regrefs
+        if isinstance(refs, RegRef):
+            refs = [refs]
+
+        if any([not r.active for r in refs]):
+            # todo allow this if the regref already has a measurement result in it.
+            # Maybe we want to delete a mode after measurement to save comp effort.
+            raise ValueError('Trying to use inactive RegRefs.')
+
+        #: list[RegRef]: register references that act as parameters for the function
+        self.regrefs = refs
+        self.func = func   #: None, function: the transformation itself, returns a scalar
+        self.func_str = func_str
+
+        if func is None and len(refs) > 1:
+            raise ValueError('Identity transformation only takes one parameter.')
+
+    def __str__(self):
+        """Print the RegRefTransform using Blackbird syntax."""
+        temp = [str(r) for r in self.regrefs]
+        rr = ', '.join(temp)
+
+        if len(temp) > 1:
+            rr = '[' + rr + ']'
+
+        if self.func is None:
+            return 'RR({})'.format(rr)
+
+        if self.func_str is None:
+            return 'RR({}, {})'.format(rr, self.func.__name__)
+
+        return 'RR({}, {})'.format(rr, self.func_str)
+
+    def __format__(self, format_spec):
+        return self.__str__()  # pragma: no cover
+
+    def __eq__(self, other):
+        "Not equal to anything."
+        return False
+
+    def evaluate(self):
+        """Evaluates the numeric value of the function if all the measurement values are available.
+
+        Returns:
+            Number: function value
+        """
+        temp = [r.val for r in self.regrefs]
+        if any(v is None for v in temp):
+            # NOTE: "if None in temp" causes an error if temp contains arrays,
```

Comment:
Is this just a note to explain the logical expression above, or is the comment about something more important (i.e., a warning about something that will lead to bugs in other parts of the code)?

Reply:
It's an explanation of the expression above.